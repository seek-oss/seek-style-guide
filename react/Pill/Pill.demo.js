import React from 'react';
import PropTypes from 'prop-types';
import { Pill, Card, Section } from 'seek-style-guide/react';
import * as sketch from './Pill.sketch';

const Container = ({ component: DemoComponent, componentProps }) => (
  <Card style={{ width: '500px' }}>
    <Section>
      <DemoComponent {...componentProps} />
      <DemoComponent {...componentProps} />
      <DemoComponent {...componentProps} />
      <DemoComponent {...componentProps} />
      <DemoComponent {...componentProps} />
      <DemoComponent {...componentProps} />
    </Section>
  </Card>
);

Container.propTypes = {
  component: PropTypes.any,
  componentProps: PropTypes.object.isRequired
};

export default {
  route: '/pill',
  title: 'Pill',
  category: 'Form',
  component: Pill,
  container: Container,
  sketch,
  initialProps: {
    children: "I'm a pill",
    onClose: null
  },
  options: [
    {
      label: 'States',
      type: 'checklist',
      states: [
        {
          label: 'Close button',
          transformProps: ({ onClose, ...props }) => ({
            ...props,
            onClose: () => console.log('On close handler called')
          })
        }
      ]
    }
  ]
};
