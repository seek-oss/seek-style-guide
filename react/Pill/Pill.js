// @flow
import styles from './Pill.less';
import React, { Component } from 'react';
import classnames from 'classnames';
import Text from '../Text/Text';
import CrossIcon from '../CrossIcon/CrossIcon';
import ScreenReaderOnly from '../ScreenReaderOnly/ScreenReaderOnly';

type Props = {
  children?: React$Node,
  text?: React$Node,
  onClose?: Function,
  className?: string
};

export default class Pill extends Component<Props> {
  static displayName = 'Pill';

  props: Props

  handleClose = (event: Object) => {
    const { onClose } = this.props;
    if (onClose) {
      onClose(event);
    }
  }

  renderStaticPill() {
    const { children, text, className, ...restProps } = this.props;
    const content = children || text;

    return (
      <span className={classnames(className, styles.staticPill)} {...restProps}>
        <Text baseline={false} raw>{content}</Text>
      </span>
    );
  }

  renderInteractivePill() {
    const { children, text, onClose, className, ...restProps } = this.props;
    const content = children || text;

    return (
      <span
        className={classnames(className, styles.interactivePill)}
        {...restProps}>
        <Text baseline={false} raw>{content}</Text>
        <button
          className={styles.removeButton}
          onClick={onClose}>
          <ScreenReaderOnly>Remove item {content}</ScreenReaderOnly>
          <div className={styles.removeCircle}>
            <CrossIcon
              className={styles.removeIcon}
              svgClassName={styles.removeSvg}
            />
          </div>
        </button>
      </span>
    );
  }

  render() {
    const { onClose } = this.props;

    return onClose ?
      this.renderInteractivePill() :
      this.renderStaticPill();
  }
}
