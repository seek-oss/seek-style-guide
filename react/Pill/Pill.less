@import (reference) "~seek-style-guide/theme";

@staticTagHeight: @row-height * 5;
@interactiveTagHeight: @row-height * 8;
@removeCircleDiameter: @row-height * 4;

.staticPill {
  .rawText(@interaction-type-scale);

  align-items: center;
  display: inline-flex;
  color: @sk-charcoal;
  background: @sk-gray-light;
  border-radius: @staticTagHeight / 2;
  height: @staticTagHeight;
  line-height: @staticTagHeight;
  padding: 0 @field-gutter-width;
  margin: @row-height 10px @row-height 0;
}

.interactivePill {
  align-items: center;
  display: inline-flex;
  background: @sk-gray-light;
  border-radius: @interactiveTagHeight / 2;
  height: @interactiveTagHeight;
  line-height: @interactiveTagHeight;
  padding: 0 0 0 (@interactiveTagHeight / 2);
  margin: @row-height 10px @row-height 0;
}

.removeButton {
  margin: 0;
  padding: 0;
  height: @interactiveTagHeight - 2;
  width: @interactiveTagHeight - 2;
  color: @sk-white;
  background: transparent;
  appearance: none;
  box-shadow: none;
  border: none;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;

  &:focus {
    outline: none;

    .removeCircle {
      .focus();

      border-radius: @removeCircleDiameter / 2;
    }
  }

  &:hover,
  &:active {
    .removeCircle {
      background-color: @sk-form-accent;
    }
  }

  &:active {
    .removeCircle {
      transform: scale(0.85);
    }
  }
}

.removeCircle {
  background-color: @sk-mid-gray;
  height: @removeCircleDiameter;
  width: @removeCircleDiameter;
  border-radius: @removeCircleDiameter / 2;
  display: flex;
  align-items: center;
  justify-content: center;
  transition: @spring-transition;
  line-height: normal;
}

.removeIcon {
  height: @removeCircleDiameter / 2;
}

.removeSvg {
  height: 100%;
}
