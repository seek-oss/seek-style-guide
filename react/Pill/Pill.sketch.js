import React from 'react';
import Pill from './Pill';
import SketchFieldContainer from '../private/SketchFieldContainer/SketchFieldContainer';

const noop = () => {};

export const symbols = {
  'Pill/Static': (
    <SketchFieldContainer>
      <Pill>Pill text</Pill>
    </SketchFieldContainer>
  ),
  'Pill/Interactive': (
    <SketchFieldContainer>
      <Pill onClose={noop}>Pill text</Pill>
    </SketchFieldContainer>
  )
};
