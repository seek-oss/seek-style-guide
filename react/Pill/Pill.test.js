import React from 'react';
import { shallow } from 'enzyme';
import Pill from './Pill';

const renderPill = props => shallow(<Pill children="I am a pill" {...props} />);

describe('Pill:', () => {
  it('should render a static pill', () => {
    const pill = renderPill();
    expect(pill).toMatchSnapshot();
  });

  it('should render a child component', () => {
    const ChildComponent = () => <div>Hello</div>;
    const pill = renderPill({ children: <ChildComponent /> });
    expect(pill).toMatchSnapshot();
  });

  it('should support the legacy "text" prop', () => {
    const ChildComponent = () => <div>Hello</div>;
    const pill = renderPill({ children: null, text: <ChildComponent /> });
    expect(pill).toMatchSnapshot();
  });

  it('should render an interactive pill', () => {
    const pill = renderPill({ onClose: () => {} });
    expect(pill).toMatchSnapshot();
  });

  it('should render an interactive pill with the legacy "text" prop', () => {
    const pill = renderPill({
      onClose: () => {},
      children: null,
      text: 'I am a pill'
    });
    expect(pill).toMatchSnapshot();
  });

  it('should call a passed in onClose function when the button fires a click event', () => {
    const onClose = jest.fn();
    const alert = renderPill({ onClose });
    const event = { someEventData: true };

    alert.find('button').simulate('click', event);
    expect(onClose).toBeCalledWith(event);
  });

  it('should pass through additional props', () => {
    const alert = renderPill({ 'custom-prop': true });
    expect(alert).toMatchSnapshot();
  });

  it('should render an additional class for the static pill', () => {
    const alert = renderPill({ className: 'test-class' });
    expect(alert).toMatchSnapshot();
  });

  it('should render an additional class for the interactive pill', () => {
    const alert = renderPill({
      className: 'test-class',
      onClose: () => {}
    });
    expect(alert).toMatchSnapshot();
  });
});
