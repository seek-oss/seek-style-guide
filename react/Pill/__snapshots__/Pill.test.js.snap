// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pill: should pass through additional props 1`] = `
<span
  className="Pill__staticPill"
  custom-prop={true}
>
  <Text
    baseline={false}
    raw={true}
  >
    I am a pill
  </Text>
</span>
`;

exports[`Pill: should render a child component 1`] = `
<span
  className="Pill__staticPill"
>
  <Text
    baseline={false}
    raw={true}
  >
    <ChildComponent />
  </Text>
</span>
`;

exports[`Pill: should render a static pill 1`] = `
<span
  className="Pill__staticPill"
>
  <Text
    baseline={false}
    raw={true}
  >
    I am a pill
  </Text>
</span>
`;

exports[`Pill: should render an additional class for the interactive pill 1`] = `
<span
  className="test-class Pill__interactivePill"
>
  <Text
    baseline={false}
    raw={true}
  >
    I am a pill
  </Text>
  <button
    className="Pill__removeButton"
    onClick={[Function]}
  >
    <ScreenReaderOnly
      component="span"
    >
      Remove item 
      I am a pill
    </ScreenReaderOnly>
    <div
      className="Pill__removeCircle"
    >
      <CrossIcon
        className="Pill__removeIcon"
        svgClassName="Pill__removeSvg"
      />
    </div>
  </button>
</span>
`;

exports[`Pill: should render an additional class for the static pill 1`] = `
<span
  className="test-class Pill__staticPill"
>
  <Text
    baseline={false}
    raw={true}
  >
    I am a pill
  </Text>
</span>
`;

exports[`Pill: should render an interactive pill 1`] = `
<span
  className="Pill__interactivePill"
>
  <Text
    baseline={false}
    raw={true}
  >
    I am a pill
  </Text>
  <button
    className="Pill__removeButton"
    onClick={[Function]}
  >
    <ScreenReaderOnly
      component="span"
    >
      Remove item 
      I am a pill
    </ScreenReaderOnly>
    <div
      className="Pill__removeCircle"
    >
      <CrossIcon
        className="Pill__removeIcon"
        svgClassName="Pill__removeSvg"
      />
    </div>
  </button>
</span>
`;

exports[`Pill: should render an interactive pill with the legacy "text" prop 1`] = `
<span
  className="Pill__interactivePill"
>
  <Text
    baseline={false}
    raw={true}
  >
    I am a pill
  </Text>
  <button
    className="Pill__removeButton"
    onClick={[Function]}
  >
    <ScreenReaderOnly
      component="span"
    >
      Remove item 
      I am a pill
    </ScreenReaderOnly>
    <div
      className="Pill__removeCircle"
    >
      <CrossIcon
        className="Pill__removeIcon"
        svgClassName="Pill__removeSvg"
      />
    </div>
  </button>
</span>
`;

exports[`Pill: should support the legacy "text" prop 1`] = `
<span
  className="Pill__staticPill"
>
  <Text
    baseline={false}
    raw={true}
  >
    <ChildComponent />
  </Text>
</span>
`;
